SMDS_3/include/CGAL/tetrahedron_soup_to_triangulation_3.h

```diff
@@ -0,0 +1,199 @@
+  * in `points`.
+  * It must
+  * describe a non self-intersecting set of tetrahedra, that cover the convex hull of the
+  * corresponding point set. The tetrahedra must form a valid triangulation with each
+  * pair of neighboring cells sharing exactly one triangle. Combinatorial validity and
+  * validity of the geometric embedding are required.
+  *
+  * @param np an optional sequence of \ref bgl_namedparameters "Named Parameters" among the ones listed below
+  *
+  * \cgalNamedParamsBegin
+  *   \cgalParamNBegin{surface_facets}
+  *     \cgalParamDescription{each element in the range describes a surface facet using the indices of points
+  *       in `points` (indices 0 to 2), and the associated `Surface_patch_index` (index 3)}
+  *     \cgalParamType{a class model of `AssociativeContainer`
+  *                    whose key type is model of `RandomAccessContainer` containing `int`
+  *                    and mapped type is `Tr::Cell::Surface_patch_index`}
+  *     \cgalParamDefault{an empty `boost::unordered_map<std::array<int, 3>, typename Tr::Cell::Surface_patch_index>`}
+  *     \cgalParamExtra{to avoid copies of large data sets, this parameter can be passed using `std::cref`}
+  *   \cgalParamNEnd
+  *   \cgalParamNBegin{subdomain_indices}
+  *     \cgalParamDescription{each element in the range gives the
+  *       `Triangulation::Cell::Subdomain_index` corresponding to the tetrahedron (cell)
+  *        of same index in `tets`}
+  *     \cgalParamType{a class model of `RandomAccessContainer` whose value type
+  *                    is `Triangulation::Cell::Subdomain_index`}
+  *     \cgalParamDefault{each finite cell of the output triangulation is
+  *                    set to have `1` as `Subdomain_index`}
+  *     \cgalParamExtra{to avoid copies of large data sets, this parameter can be passed using `std::cref`}
+  *   \cgalParamNEnd
+  * \cgalNamedParamsEnd
+  *
+  * @returns the 3D triangulation built from parameters
+  *
+  * @pre `points` contains each point only once
+  * @post the output triangulation must be a triangulation of the convex hull of `points`
+  * @post `is_valid()` returns `true` for the returned triangulation
+  *
+  * @sa `CGAL::Polygon_mesh_processing::polygon_soup_to_polygon_mesh()`
+  */
+  template<typename Triangulation,
+           typename PointRange,
+           typename TetrahedronRange,
+           typename NamedParameters>
+  Triangulation tetrahedron_soup_to_triangulation_3(const PointRange& points,
+                                                    const TetrahedronRange& tets,
+                                                    const NamedParameters& np)
+  {
+    using parameters::choose_parameter;
+    using parameters::get_parameter;
+    using parameters::get_parameter_reference;
+
+    Triangulation tr;
+    boost::unordered_map<std::array<int, 3>,
+             typename Triangulation::Cell::Surface_patch_index> empty_map;
+    const auto& facets = choose_parameter(
```

Comment:
The return type must be deduced like in https://github.com/CGAL/cgal/blob/master/STL_Extension/test/STL_Extension/test_cgal_named_params.cpp#L62, otherwise you get a dangling reference if the user does not actually pass a reference.